keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+    - [Code structure](#code-structure)
+    - [Timeframe and Implementation Order](#timeframe-and-implementation-order)
+    - [OWNERS file for new packages](#owners-file-for-new-packages)
+
+## Abstract
+
+The proposal is about preparing the Kubernetes core package structure in a way that all core component can share common code around
+
+- ComponentConfig implementation
+- flag and command handling
+- HTTPS serving
+- delegated authn/z
+- logging.
+
+Today this code is spread over the k8s.io/kubernetes repository, staging repository or pieces of code are in locations they don't belong to (example: k8s.io/apiserver/pkg/util/logs is the for general logging, totally independent of API servers). We miss a repository far enough in the dependency hierarchy for code that is or should be common among core Kubernetes component (neither k8s.io/apiserver, k8s.io/apimachinery or k8s.io/client-go are right for that).
+
+### History and Motivation
+
+By this time in the Kubernetes development, we know pretty well how we want a Kubernetes component to work, function, and look. But achieving this requires a fair amount of more or less advanced code. As we scale the ecosystem, and evolve Kubernetes to work more as a kernel, it's increasingly important to make writing extensions and custom Kubernetes-aware components relatively easy. As it stands today, this is anything but straightforward. In fact, even the in-core components diverge in terms of configurability (Can it be declaratively configured? Do flag names follow a consistent pattern? Are configuration sources consistently merged?), common functionality (Does it support the common "/version," "/healthz," "/configz," "/pprof," and "/metrics" endpoints? Does it utilize Kubernetes' authentication/authorization mechanisms? Does it write logs in a consistent manner? Does it handle signals as others do?), and testability (Do the internal configuration structs set up correctly to conform with the Kubernetes API machinery, and have roundtrip, defaulting, validation unit tests in place? Does it merge flags and the config file correctly? Is the logging mechanism set up in a testable manner? Can it be verified that the HTTP server has the standard endpoints registered and working? Can it be verified that authentication and authorization is set up correctly?).
+
+This document proposes to create a new Kubernetes staging repository with minimal dependencies (_k8s.io/apimachinery_, _k8s.io/client-go_, and _k8s.io/api_) and good documentation on how to write a Kubernetes-aware component that follows best practices. The code and best practices in this repo would be used by all the core components as well. Unifying the core components would be great progress in terms of the internal code structure, capabilities, and test coverage. Most significantly, this would lead to an adoption of ComponentConfig for all internal components as both a "side effect" and a desired outcome, which is long time overdue.
```

Comment:
Are we considering a working group at all? There's a lot of detail and nuance around the items discussed in this section, so it might be useful to have the structure that a working group provides.

Reply:
WG seems like a good idea to me.

Reply:
Yeah, that'd be good, but I'm not sure it'd be the right thing for me to bootstrap a WG while doing my military service :/. In other words, I can attend (some) meetings and provide async coding/reviews, but probably not do all the bureaucracy of creating a WG at this time

@mtaufen if you want to bootstrap I'm happy to help with what I can though :wink:

Reply:
I will also help in this endeavor.

Reply:
ok, thanks. I'll get back to you on what we need to get done later :+1: